### PR TITLE
nbqa: add passthru.withTools

### PR DESCRIPTION
**Pull request by dotlambda**
## Description of changes
You can now do fancy things like

    (nbqa.override { python3 = python311; }).withTools (ps: [ ps.black ])

I didn't apply nixfmt for now in order to keep the diff minimal. Any suggestion for making both happen?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -6605,7 +6605,7 @@ with pkgs;
 
   bbin = callPackage ../development/tools/bbin { };
 
-  blacken-docs = callPackage ../tools/misc/blacken-docs { };
+  blacken-docs = with python3Packages; toPythonApplication blacken-docs;
```
Comment (Pandapip1) [inline pkgs/top-level/all-packages.nix:6608]:
```suggestion
  blacken-docs = toPythonApplication blacken-docs;
```

**Comment by l0b0** [comment]
> I didn't apply nixfmt for now in order to keep the diff minimal. Any suggestion for making both happen?

A separate commit applying `nixfmt` (and only that) before the commit with the functional change would make it easy to review the functional change separately.

**Comment by dotlambda** [comment]
> > I didn't apply nixfmt for now in order to keep the diff minimal. Any suggestion for making both happen?
> 
> A separate commit applying `nixfmt` (and only that) before the commit with the functional change would make it easy to review the functional change separately.

Do you mean after? The problem is that introducing `let ... in` would indent everything by two spaces.

**Comment by l0b0** [comment]
> > > I didn't apply nixfmt for now in order to keep the diff minimal. Any suggestion for making both happen?
> > 
> > 
> > A separate commit applying `nixfmt` (and only that) before the commit with the functional change would make it easy to review the functional change separately.
> 
> Do you mean after? The problem is that introducing `let ... in` would indent everything by two spaces.

No, before. It's easy to ignore space changes when reviewing.

**Comment by dotlambda** [comment]
@l0b0 What you're saying doesn't make sense. The functional change (introducing `let`) is exactly what requires reformatting.

**Comment by l0b0** [comment]
> @l0b0 What you're saying doesn't make sense. The functional change (introducing `let`) is exactly what requires reformatting.

`pkgs/tools/misc/blacken-docs/default.nix` should be reformatted before moving it. But you're right that `pkgs/by-name/nb/nbqa/package.nix` doesn't need reformatting before the functional change. In that case, the reformatting should be done as part of the functional change commit, so that each commit passes linting.

**Comment by dotlambda** [comment]
> `pkgs/tools/misc/blacken-docs/default.nix` should be reformatted before moving it

No point because git correctly identifies the file as being moved.

**Comment by dotlambda** [comment]
> In that case, the reformatting should be done as part of the functional change commit, so that each commit passes linting.

That doesn't answer my question of how to do it with a minimal diff.

**Comment by l0b0** [comment]
> > In that case, the reformatting should be done as part of the functional change commit, so that each commit passes linting.
> 
> That doesn't answer my question of how to do it with a minimal diff.

Making each commit valid is more important than having the minimal diff.

**Comment by l0b0** [comment]
> > `pkgs/tools/misc/blacken-docs/default.nix` should be reformatted before moving it
> 
> No point because git correctly identifies the file as being moved.

If you create a commit which only moves the file then that commit won't pass linting.

**Comment by dotlambda** [comment]
> If you create a commit which only moves the file then that commit won't pass linting.

There is no such commit and there won't be. Please stop.

**Comment by l0b0** [comment]
> > If you create a commit which only moves the file then that commit won't pass linting.
> 
> There is no such commit and there won't be. Please stop.

When you said:

> > `pkgs/tools/misc/blacken-docs/default.nix` should be reformatted before moving it
> 
> No point because git correctly identifies the file as being moved.

That's not the only thing you need to do when moving a file. You also need to change anywhere that file is referenced. That's a functional change, which means the files involved must be `nixfmt`ed. By doing that in a separate commit *before* moving the file, you could keep each commit as simple as possible.

**Comment by dotlambda** [comment]
Alright, enough. I blocked you.

**Comment by RuRo** [comment]
@dotlambda did you just block the current maintainer of this package for trying to help you (even though you originally explicitly asked for input about that)? Or am I horribly misreading the situation?

**Comment by dotlambda** [comment]
They were not helpful at all. They suggested I turn the commit for blacken-docs into two for no reason and didn't comment on how to improve the withTools commit.

File: pkgs/by-name/nb/nbqa/package.nix [hunk 2]
```diff
@@ -7,93 +7,111 @@
   ruff,
 
   # tests
-  versionCheckHook
+  versionCheckHook,
```
Comment (dotlambda) [inline pkgs/by-name/nb/nbqa/package.nix:10]: Comma should be added in the previous commit.

**Pull request merged**